shaders/src/fog.fs

```diff
@@ -2,43 +2,49 @@
-        float h = dot(view, up);
-        h = abs(h) > 0.01 ? h : 0.01;
-        float falloff = max(frameUniforms.fogHeightFalloff * h, -127.0);
-
-        // fog density at shaded fragment w.r.t. camera's altitude
-        float densityFalloff = density * (1.0 - exp2(-falloff)) / falloff;
-
-        // take fog start distance into account
-        float lineIntegral = densityFalloff * max(cameraToWorldDistanceClamped - frameUniforms.fogStart, 0.0);
-        float fogFactor = max(clamp(exp2(-lineIntegral), 0.0, 1.0), 1.0 - frameUniforms.fogMaxOpacity);
-
-        vec3 foggedColor = frameUniforms.fogColor * (1.0 - fogFactor);
-        if (frameUniforms.fogInscatteringSize > 0.0) {
-            vec3 lightDirectionNorm = -normalize(frameUniforms.lightDirection);
-            vec3 cameraToWorldUnit = view * (1.0 / cameraToWorldDistance);
-            float lightInscatteringFactor = max(dot(cameraToWorldUnit, lightDirectionNorm), 0.0);
-            vec3 lightInscattering = frameUniforms.lightColorIntensity.rgb * frameUniforms.lightColorIntensity.w * pow(lightInscatteringFactor, frameUniforms.fogInscatteringSize);
-            float lightInscatteringIntegral = densityFalloff * max(cameraToWorldDistanceClamped - frameUniforms.fogInscatteringStart, 0.0);
-            float inscatteringFactor = clamp(exp2(-lightInscatteringIntegral), 0.0, 1.0);
-            foggedColor += lightInscattering * (1.0 - inscatteringFactor);
+        float A = frameUniforms.fogDensity;
+        float B = frameUniforms.fogHeightFalloff;
+
+        float d = length(view);
+        float h = abs(view.y) < FLT_EPS ? FLT_EPS : view.y;   // avoid divide-by-zero
+
+        float fogIntegralFunctionOfDistance = A * (1.0 - exp2(-B * h)) / h;
+        float fogIntegral = fogIntegralFunctionOfDistance * max(d - frameUniforms.fogStart, 0.0);
+
+        // remap fogIntegral to an opacity between 0 and 1. saturate(fogIntegral) is too harsh.
+        float fogOpacity = max(1.0 - exp2(-fogIntegral), 0.0);
+        //float fogOpacity = clamp(fogIntegral, 0.0, 1.0);
+
+        // don't go above requested max opacity
+        fogOpacity = min(fogOpacity, frameUniforms.fogMaxOpacity);
+
+        // compute fog color
+        vec3 fogColor = frameUniforms.fogColor;
+
+        if (frameUniforms.fogColorFromIbl > 0.0) {
+            // get fog color from envmap
+            float lod = frameUniforms.iblRoughnessOneLevel;
+            fogColor = textureLod(light_iblSpecular, view, lod).rgb * frameUniforms.iblLuminance;
+        }
+
+        fogColor *= fogOpacity;
+        if (frameUniforms.fogInscatteringSize >= 0.0) {
+            // compute a new line-integral for a different start distance
+            float inscatteringIntegral = fogIntegralFunctionOfDistance * max(d - frameUniforms.fogInscatteringStart, 0.0);
+            float inscatteringOpacity = max(1.0 - exp2(-inscatteringIntegral), 0.0);
+            //float inscatteringOpacity = clamp(inscatteringIntegral, 0.0, 1.0);
```

Comment:
Remove?

Reply:
I wanted to keep it for now until I'm really sure. After reading more articles about this, I still don't see anyone doing this (AFAIK), which is strange.

Reply:
It seems easy to do and worth it. It's an extra feature at super low cost and could be useful to simulate time of day in some cases.